### PR TITLE
Add pre-processing flag and functions applicable to fMOST datastes

### PR DESCRIPTION
**Pull request by noisysky**
Add `--pre-processing` command line option, which accepts strings such as 'fmost', 'none', 'default'.
Add pre-processing functions specifically for fMOST datasets, which remove horizontal stripes and background.

### CODE REVIEW

**Comment by adamltyson** [comment]
Hey @noisysky, thanks for this PR. A couple of things:
* OpenCV can sometimes cause installation issues on different OSs, is there a version of the function you need in numpy, scipy, scikit-image etc?
* Could you add some tests for this new functionality?
* Is there some example data to test this new functionality?

**Comment by noisysky** [comment]
@adamltyson Thanks for the review! 
- Replaced OpenCV connectedComponents with skimage label and regionprops.
- Added 1 test (if it's not enough, please help me to come up with ideas for tests).
- The test data can be found here: https://download.brainlib.org/hackathon/2022_GYBS/input/fMOST/subject/. I had to convert .nii.gz files to .tif before running brainreg on them.

**Comment by adamltyson** [comment]
Hey @noisysky, sorry for the very delayed response. This looks good, I've left some comments in the review.

BTW - brainreg will accept `.nii` files, just not the compressed versions. I want to extend this, but in the context of a general purpose I/O library (see image.sc discussion - https://forum.image.sc/t/developing-a-common-framework-to-represent-anatomical-data/68860).

**Review by adamltyson — CHANGES_REQUESTED** [review]

File: brainreg/cli.py [hunk 1]
```diff
@@ -165,6 +166,23 @@ def geometry_parser(parser):
     return parser
 
 
+def preprocessing_parser(parser):
+    preprocessing_opt_parser = parser.add_argument_group(
+        "Pre-processing options"
+    )
+
+    preprocessing_opt_parser.add_argument(
```
[the remark this answers is not in the document]
Comment (noisysky) [reply brainreg/cli.py:185]: Yes, added the default value.

File: tests/tests/test_fmost_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import os
+
+import numpy as np
+from numpy.testing import assert_array_equal, assert_raises
+import tifffile
+
+from brainreg.utils.preprocess import subtract_background
+
+
+def test_background_removal():
```
[the remark this answers is not in the document]
Comment (noisysky) [reply tests/tests/test_fmost_preprocessing.py:10]: Yes, I uploaded a pair of images to the test/data/input folder, and the test you suggested runs on them.

File: brainreg/backend/niftyreg/run.py [hunk 3]
```diff
@@ -49,13 +50,18 @@ def run_niftyreg(
...
 
-    target_brain = preprocess.filter_image(target_brain)
+    target_brain = preprocess.filter_image(target_brain, preprocessing_args)
     save_nii(
         target_brain,
         atlas.resolution,
         niftyreg_paths.downsampled_filtered,
     )
 
+    imio.to_tiff(
```
[the remark this answers is not in the document]
Comment (noisysky) [reply brainreg/backend/niftyreg/run.py]: I removed it. It was there for debug and I forgot about it :)

File: brainreg/utils/preprocess.py [hunk 4]
```diff
@@ -1,19 +1,27 @@
...
     :return: The filtered brain
     :rtype: np.array
     """
     brain = brain.astype(np.float64, copy=False)
-    for i in trange(brain.shape[-1], desc="filtering", unit="plane"):
-        brain[..., i] = filter_plane(brain[..., i])
+    if preprocessing_args and preprocessing_args.preprocessing == "none":
+        pass
+    elif preprocessing_args and preprocessing_args.preprocessing == "fmost":
+        for i in trange(brain.shape[0], desc="filtering", unit="plane"):  # only coronal plane because of stripes
```
[the remark this answers is not in the document]
Comment (noisysky) [reply brainreg/utils/preprocess.py]: The brain at this point is always reoriented to the atlas orientation. With that comment I tried to justify why I am slicing images in the other dimension than the default pre-processing: in coronal plane stripes are most visible / easiest to remove. I removed the comment.

File: brainreg/paths.py [hunk 5]
```diff
@@ -14,6 +14,7 @@ def __init__(self, registration_output_folder):
 
     def make_reg_paths(self):
         self.downsampled_brain_path = self.make_reg_path("downsampled.tiff")
+        self.downsampled_filtered_brain_path = self.make_reg_path("downsampled_filtered.tiff")
```
[the remark this answers is not in the document]
Comment (noisysky) [reply brainreg/paths.py]: Removed this as well

File: brainreg/utils/preprocess.py [hunk 6]
```diff
@@ -61,3 +69,135 @@ def pseudo_flatfield(img_plane, sigma=5):
...
+    for d in range(len(points)):
+        u0, v0 = points[d]
+        mask1 = (u - u0)**2 + (v - v0)**2 <= d0
+        mask2 = (u + u0)**2 + (v + v0)**2 <= d0
+        fshift[mask1] = 0
+        fshift[mask2] = 0
+    return fshift
+
+
+def remove_stripes(img_plane, stripes_direction="h"):
```
[the remark this answers is not in the document]
Comment (noisysky) [reply brainreg/utils/preprocess.py]: It looks like all fMOST datasets that I've seen have the same stripes direction, so I only retained "horizontal stripes" part of the code.

**Comment by adamltyson** [comment]
Hey @noisysky, this is looking in pretty good shape now (I left a few more comments). A couple more things though:
* It would be great to add some more tests to at least make sure the test coverage % doesn't drop.
* When running brainreg on some fmost datasets, I saw some signal missing in the `downsample_filtered.nii` image that's used for registration. This isn't present in the raw data (or the downsampled stack). There's also a misalignment around the cerebellum, but that's present in the raw data.
Screenshot from 2022-08-10 13-10-28

Have you seen this before with the preprocessing?

**Comment by noisysky** [comment]
Hi @adamltyson, there was a major problem that few z planes got almost entirely deleted. Please give me a few days to check if the new version of the processing that I have performs better on this data.

**Comment by noisysky** [comment]
15257_no_bg_fft_for_PR
@adamltyson My newer version doesn't produce as bad artifacts as in your image above, but produces minor artifacts anyway. Not sure if I can improve it much. I abandoned doing background removal using standard methods, and moved on to RandomForest pixel classifiers for this purpose. Also, I noticed that fMOST align better without any pre-processing at all, compared to standard brainreg pre-processing, therefore it makes sense to keep the option of disabling the pre-processing.
I'll push the newest version of background removal that I have, and let you decide whether you want it.

**Comment by adamltyson** [comment]
> Also, I noticed that fMOST align better without any pre-processing at all, compared to standard brainreg pre-processing, therefore it makes sense to keep the option of disabling the pre-processing.

In that case, would you mind if I just kept the option of disabling pre-processing, and not the fMOST-specific preprocessing? Long term, I want to move away from the brainreg CLI, and move to:
* Standard - combining brainreg napari plugin with flexible preprocessing provided by other napari plugins
* Advanced users - using brainreg Python API along with custom preprocessing

**Comment by noisysky** [comment]
@adamltyson Sure! If I have a radical improvement, I'll let you know and we'll open a new PR.

**Comment by adamltyson** [comment]
Great, thanks. At some point I'll move the preprocessing into it's own module, then I'll check back to see what the current status of fmost preprocessing is.

**Review by adamltyson — APPROVED** [review]

**Pull request merged**